src/time.cpp

```diff
@@ -65,11 +127,40 @@ namespace g3 {
       return tm_snapshot;
    }
 
-   /// returns a std::string with content of time_t as localtime formatted by input format string
-   /// * format string must conform to std::put_time
-   /// This is similar to std::put_time(std::localtime(std::time_t*), time_format.c_str());
 
-   std::string localtime_formatted(const std::time_t &time_snapshot, const std::string &time_format) {
+
+   size_t fractional_cutoff(const std::string& format_buffer, size_t pos) {
+      char  ch  = (format_buffer.size() > pos + kIdentifierSize ? format_buffer.at(pos + kIdentifierSize) : '\0');
+      size_t cutOff = 0;
+      switch (ch) {
+         case '3': cutOff = kMsDigitsToCut; break;
+         case '6': cutOff = kUsDigitsToCut; break;
+         case '9': cutOff = kNsDigitsToCut; break;
+         default: cutOff = kNsDefaultDigits; break;
+      }
+      return cutOff;
+   }
+
+
+   std::string localtime_formatted(const timespec& time_snapshot, const std::string& time_format) {
+      auto format_buffer = time_format;  // copying format string to a separate buffer
+
+      // iterating through every "%f" instance in the format string
+      auto identifierExtraSize = 0;
+      for (size_t pos = 0; (pos = format_buffer.find(kIdentifier, pos)) != std::string::npos; pos += kIdentifierSize + identifierExtraSize) {
+         // figuring out whether this is nano, micro or milli identifier
+         auto cutoff = fractional_cutoff(format_buffer, pos);
+         auto value = fractional_to_string(time_snapshot.tv_nsec, cutoff);
```

Comment:
consider having the function which makes more sense
```g3::fractional_to_string(time_snapshot /*i.e. it takes the timespec*/,   FractionalType::us/ms/ns);```
it's just way more readable than the ```1000000000, 1000000, 1000``` and with one ```if-else``` switch all the nastiness is dealt with